### ISSUE TITLE
JS Interop with Option does not appear to be working

### ISSUE THREAD

**Issue by jgrund** [fable-compiler/Fable#1312]
Previously Fable appeared to compile option types by unwrapping the value for `Some` and replacing `None` with `null`.

This was nice for interop with JS APIs that expected `null` as an actual value. For instance the nodejs `stream.Readable.prototype.push` is called with `null` to signal the end of the stream.

Not sure if I'm missing something, but in the latest Fable version (1.3.7) it now appears that `None` is no longer compiled to `null`, but a wrapper at runtime, I.E. `None` compiles to:

```Some { value: null }```

Due to this, we lose the nice abstraction that you can think of JS `null` as `None` and any other value as `Some` (where a param could take the union of the type or `null`). This is similar to how JS type systems like flow model possibly null values.

Is this the expected behavior for `Option` types going forward? Is there some nice way we can still model the abstraction of `None` -> `null`?

**Comment by jgrund** [comment]
Sorry for the noise, looks like the unwrapping is working ok for the call.

It's actually a lambda that takes tupled args getting wrapped into an array at call time.

Because of that, the first arg isn't null but an array.

**Comment by jgrund** [comment]
Just to be clear, an abstract member like:

```abstract transform: Option<'a -> string -> (Error option * 'b option -> unit) -> unit> with get, set```

where ```(Error option * 'b option -> unit)``` is a callback.

Is then used like:

```
transformOpts.transform <- Some (fun chunk _ cb ->
        cb(None, Some (chunk + "bar"))
    )
```

This is compiled to:

```
    transformOpts.transform = function (chunk, _arg2, cb) {
        cb([null, chunk + "bar"]);
    };
```

Where I was expecting the call to look more like:

```
    transformOpts.transform = function (chunk, _arg2, cb) {
        cb(null, chunk + "bar");
    };
```

Unsure if that's expected behavior or not.

**Comment by Zaid-Ajaj** [comment]
Shouldn't you use `(Error option -> 'b option -> unit)` instead of `(Error option * 'b option -> unit)`? Because Fable has that optimization thingy that uncurries functions, this optimization is specially made for interop scenario's

**Comment by jgrund** [comment]
Yes, I should have (and ended up) using the curried form.

**Comment by alfonsogarciacaro** [comment]
#### About options

Yes, options were erased originally in Fable but this was causing problems with nested options and `unit option`, so in those cases [Fable now wraps the values](https://github.com/fable-compiler/Fable/blob/7229be70aa0a5acef9413c720cc7708a593bce34/src/dotnet/Fable.Compiler/FSharp2Fable.fs#L139-L152). It also does it for generic types, as it cannot know if the generic will become a nested option or `unit`.

Interop with JS should remain more or less the same because options are still erased in most cases, but for generics it may be tricky. In this case it's OK because the callback is also Fable code so it will recognize the wrapping, but it can be a problem if you send a wrapped option to JS because the signature is `'a option`. I'm not sure what to do here: Don't wrap generic options? (Though that would make [this test](https://github.com/fable-compiler/Fable/blob/7229be70aa0a5acef9413c720cc7708a593bce34/tests/Main/OptionTests.fs#L201-L205) fail.) Use an attribute to mark options that shouldn't be wrapped? (But that would make bindings more cumbersome.)

#### About curried/tupled functions

Yes, this is one most confusing parts when working with F# AST. In theory, there are no functions with multiple arguments in F#, only curried `'a->'b->'c` and tupled-argument `('a*'b)->'c` functions. However, for **methods** (including type and module methods) **both** curried and tupled-arg functions are represented as multi-arg functions (same as `(a, b) => c` in JS) in the AST so this is not an issue. The problem comes when the functions are used as values, where the F# compiler enforces the semantics and transform the functions in actual curried `a => b => c` or tupled-arg `([a,b]) => c` (Fable represents tuples as arrays) functions.

Tupled-arg functions cannot be touched (sometimes you do need to pass a tuple as a single parameter) but since v1.0.0 Fable converts curried functions into multi-arg functions in JS (using the [dynamic CurriedLambda](https://github.com/fable-compiler/Fable/blob/7229be70aa0a5acef9413c720cc7708a593bce34/src/js/fable-core/CurriedLambda.ts) when the arity cannot be inferred at compile time) for performance and also for easier interop with JS.

So in conclusion, when representing a function value (usually a callback) in Fable bindings, you should use curried functions (as it will be converted to multi-arg functions by Fable) or `System.Func` (as it was done before).
